### ISSUE TITLE
Infinite mount loop in automounter

### ISSUE THREAD

**Issue by alexcern** [cern-eos/eosxd-csi#19]
Since a few weeks I've observed several cases where an app would block on EOS access. It's not clear when/how exactly the problem started, I've seen it for the first time while updating staging clusters to OKD 4.15 and it's since been observed a few times on PaaS and WebEOS production clusters.

No clue yet as to how to reproduce the situation.

I observed the following conditions:

- a user app is trying to access an EOS path, but this access seems to block indefinitely (though it is interruptible)
- the block is in the automounter, which is struggling to mount the requested path and retries every 5s. See `automount` logs below. Each attempt is successful in automount logs but the path never becomes available to the client app.
- presumably the mount operation is misbehaving because we are in a situation where we have (old-ish) `eosxd` fuse processes for the desired mount, but not the corresponding mount point (my suspicion is that the eosxd mount returns successfully but without doing anything because it sees the existing eosxd fuse processes)

It is not clear how we end up in this situation where the mount point is gone while the eosxd fuse processes are still running.

Killing the eosxd processes that have no corresponding mount point in the output of `findmnt -t fuse` solves the issue: the next mount attempt by automount works properly and application IO unblocks.

Further info:

The `automount` container shows mount attempts every 5s:

```
I0723 15:14:53.770105 9438 automount.go:136] automount[9570]: attempting to mount entry /eos/project
I0723 15:14:53.770180 9438 automount.go:136] automount[9570]: set_tsd_user_vars: failed to get passwd info from getpwuid_r
I0723 15:14:58.833278 9438 automount.go:136] automount[9570]: mounted /eos/project
I0723 15:14:58.833628 9438 automount.go:136] automount[9570]: attempting to mount entry /eos/project
I0723 15:14:58.833821 9438 automount.go:136] automount[9570]: set_tsd_user_vars: failed to get passwd info from getpwuid_r
I0723 15:15:03.892187 9438 automount.go:136] automount[9570]: mounted /eos/project
I0723 15:15:03.892876 9438 automount.go:136] automount[9570]: attempting to mount entry /eos/project
I0723 15:15:03.893171 9438 automount.go:136] automount[9570]: set_tsd_user_vars: failed to get passwd info from getpwuid_r
I0723 15:15:08.954659 9438 automount.go:136] automount[9570]: mounted /eos/project
I0723 15:15:08.957246 9438 automount.go:136] automount[9570]: attempting to mount entry /eos/project
I0723 15:15:08.957259 9438 automount.go:136] automount[9570]: set_tsd_user_vars: failed to get passwd info from getpwuid_r
I0723 15:15:14.014015 9438 automount.go:136] automount[9570]: mounted /eos/project
I0723 15:15:14.014091 9438 automount.go:136] automount[9570]: attempting to mount entry /eos/project
I0723 15:15:14.014386 9438 automount.go:136] automount[9570]: set_tsd_user_vars: failed to get passwd info from getpwuid_r
I0723 15:15:19.077611 9438 automount.go:136] automount[9570]: mounted /eos/project
I0723 15:15:19.077858 9438 automount.go:136] automount[9570]: attempting to mount entry /eos/project
I0723 15:15:19.078064 9438 automount.go:136] automount[9570]: set_tsd_user_vars: failed to get passwd info from getpwuid_r
I0723 15:15:24.142322 9438 automount.go:136] automount[9570]: mounted /eos/project
I0723 15:15:24.142427 9438 automount.go:136] automount[9570]: attempting to mount entry /eos/project
I0723 15:15:24.144235 9438 automount.go:136] automount[9570]: set_tsd_user_vars: failed to get passwd info from getpwuid_r
I0723 15:15:29.206128 9438 automount.go:136] automount[9570]: mounted /eos/project
I0723 15:15:29.206289 9438 automount.go:136] automount[9570]: attempting to mount entry /eos/project
I0723 15:15:29.206502 9438 automount.go:136] automount[9570]: set_tsd_user_vars: failed to get passwd info from getpwuid_r
I0723 15:15:34.261543 9438 automount.go:136] automount[9570]: mounted /eos/project
I0723 15:15:34.261711 9438 automount.go:136] automount[9570]: attempting to mount entry /eos/project
I0723 15:15:34.261880 9438 automount.go:136] automount[9570]: set_tsd_user_vars: failed to get passwd info from getpwuid_r
```

I'll collect more data the next time I see it happening.

**Comment by alexcern** [comment]
@1602077 I opened this issue for reference, at this point it seems best to wait for more occurrences and better forensics before taking action.

But in case this keeps happening and we can't figure out why, we should be able to work around it by having the `mountreconciler` kill eosxd processes without a mount point (mirroring the current behavior of unmounting mount points without an eosxd process https://github.com/cern-eos/eosxd-csi/blob/main/internal/eosxd/mountreconcile/reconcile.go#L206)

**Comment by alexcern** [comment]
This may be what caused the old-eosxd-process-without-mountpoint situation: (from `/var/log/eos/fusex/fuse.project.log` on the `automount` container)

```
Logging: suspended 1 running 0 in q 0
Logging: suspended 0 running 1 in q 0
240723 11:20:26 t=1721726426.735041 f=DumpStatistic    l=WARN  tid=00007f4a30e00640 s=eosfuse:2720             HB (heartbeat) has not started!
240723 11:20:26 t=1721726426.735753 f=mdcallback       l=WARN  tid=00007f4a2ac00640 s=md:2814                  MGM asked us to set our heartbeat interval to 10 seconds, enable dentry
-messaging, enable writesizeflush, accepts appname, accepts mdquery versions hidden and server-version=5.2.24::1
240723 11:20:26 t=1721726426.834595 f=run              l=WARN  tid=00007f4a46307840 s=eosfuse:2075             ********************************************************************************
240723 11:20:26 t=1721726426.834629 f=run              l=WARN  tid=00007f4a46307840 s=eosfuse:2076             eosxd started version 5.2.22 - FUSE protocol version 28
240723 11:20:26 t=1721726426.834633 f=run              l=WARN  tid=00007f4a46307840 s=eosfuse:2078             eos-instance-url       := eosproject-fuse.cern.ch
240723 11:20:26 t=1721726426.834640 f=run              l=WARN  tid=00007f4a46307840 s=eosfuse:2079             thread-pool            := custom
240723 11:20:26 t=1721726426.834642 f=run              l=WARN  tid=00007f4a46307840 s=eosfuse:2081             zmq-connection         := tcp://eosproject-fuse.cern.ch:1100
240723 11:20:26 t=1721726426.834647 f=run              l=WARN  tid=00007f4a46307840 s=eosfuse:2082             zmq-identity           := fuse://eosxd@paas-standard-avz-a-j4nm5//cc1ee
29c-48d4-11ef-bea3-fa163e850c0c:13486
240723 11:20:26 t=1721726426.834650 f=run              l=WARN  tid=00007f4a46307840 s=eosfuse:2083             fd-limit               := 524288
240723 11:20:26 t=1721726426.834676 f=run              l=WARN  tid=00007f4a46307840 s=eosfuse:2086             sss-keytabfile         := /etc/eos.keytab
240723 11:20:26 t=1721726426.834679 f=run              l=WARN  tid=00007f4a46307840 s=eosfuse:2090             ztn token              := enabled
240723 11:20:26 t=1721726426.834684 f=run              l=WARN  tid=00007f4a46307840 s=eosfuse:2093             options                := backtrace=1 md-cache:1 md-enoent:0.00 md-time
out:86400.00 md-put-timeout:120.00 data-cache:1 rename-sync:1 rmdir-sync:0 flush:0 flush-w-open:1 flush-w-open-sz:262144 flush-w-umount:120 locking:1 no-fsync:.db,.db-journal,.sqlite
,.sqlite-journal,.db3,.db3-journal,.o, flush-nowait-exec:/tar,/touch, ol-mode:000 show-tree-size:0 hide-versions:1 protect-symlink-loops:0 core-affinity:0 no-xattr:0 no-eos-xattr-lis
ting: 0 no-link:0 nocache-graceperiod:5 rm-rf-protect-level=0 rm-rf-bulk=0 t(lease)=300 t(size-flush)=10 submounts=0 ino(in-mem)=16384 flock:1
240723 11:20:26 t=1721726426.834695 f=run              l=WARN  tid=00007f4a46307840 s=eosfuse:2126             cache                  := rh-type:dynamic rh-nom:262144 rh-max:2097152 
rh-blocks:16 rh-sparse-ratio:0.0 max-rh-buffer=134217728 max-wr-buffer=134217728 tot-size=1048576000 tot-ino=65536 jc-size=0 jc-ino=0 dc-loc:/var/cache/eos/fusex/cache/project jc-loc
:/var/cache/eos/fusex/cache/project clean-thrs:85.000000%
240723 11:20:26 t=1721726426.834702 f=run              l=WARN  tid=00007f4a46307840 s=eosfuse:2141             read-recovery          := enabled:1 ropen:1 ropen-noserv:1 ropen-noserv
-window:15
240723 11:20:26 t=1721726426.834706 f=run              l=WARN  tid=00007f4a46307840 s=eosfuse:2146             write-recovery         := enabled:1 wopen:1 wopen-noserv:1 wopen-noserv
-window:15
240723 11:20:26 t=1721726426.834710 f=run              l=WARN  tid=00007f4a46307840 s=eosfuse:2151             file-inlining          := emabled:0 max-size=0 compressor=none
240723 11:20:26 t=1721726426.834712 f=run              l=WARN  tid=00007f4a46307840 s=eosfuse:2155             fuzzing                := open-async-submit:0(fatal:0) open-async-retur
n:0(fatal:0) read-async-return:0
240723 11:20:26 t=1721726426.834740 f=run              l=WARN  tid=00007f4a46307840 s=eosfuse:2176             xrdcl-options          := TimeoutResolution:1 ConnectionWindow:10 Conne
ctionRetry:0 StreamErrorWindow:120 RequestTimeout:60 StreamTimeout:120 RedirectLimit:2  log-level='' fusex-chunk-timeout=300
240723 11:20:26 t=1721726426.836967 f=lookupNonLocalJail l=ALERT tid=00007f4a26e00640 s=SecurityChecker:198      Failed to openat next child
240723 11:20:27 t=1721726427.735906 f=DumpStatistic    l=WARN  tid=00007f4a30e00640 s=eosfuse:2724             HB (heartbeat) has started!
240723 11:20:29 t=1721726429.053115 f=lookupNonLocalJail l=ALERT tid=00007f4a26e00640 s=SecurityChecker:198      Failed to openat next child
240723 11:20:29 t=1721726429.053203 f=lookupNonLocalJail l=ALERT tid=00007f4a26e00640 s=SecurityChecker:212      Failed to openat file
                 ---- high rate error messages suppressed ----
240723 11:25:04 t=1721726704.841008 f=lookupNonLocalJail l=ALERT tid=00007f4a28000640 s=SecurityChecker:198      Failed to openat next child
                 ---- high rate error messages suppressed ----
                 ---- high rate error messages suppressed ----
240723 11:25:06 t=1721726706.860515 f=lookupNonLocalJail l=ALERT tid=00007f4a28000640 s=SecurityChecker:212      Failed to openat file
240723 11:35:03 t=1721727303.276879 f=lookupNonLocalJail l=ALERT tid=00007f4a26e00640 s=SecurityChecker:198      Failed to openat next child
                 ---- high rate error messages suppressed ----
240723 11:35:05 t=1721727305.753049 f=lookupNonLocalJail l=ALERT tid=00007f4a26e00640 s=SecurityChecker:212      Failed to openat file
                 ---- high rate error messages suppressed ----
fusermount: failed to unmount /eos/project: Invalid argument
# umounthandler: executing fusermount -u -z /eos/project
# umounthandler: sighandler received signal 15 - emitting signal 15 again
fusermount: failed to unmount /eos/project: Invalid argument
Logging: suspended 1 running 0 in q 0
Logging: suspended 0 running 1 in q 0
```

Notice the error at unmount

**Comment by smithdh** [comment]
Hi @alexcern 
I'm not certain how the eosxd unmount failed here. But I did notice in the log snippet it mentions eosxd 5.2.22. In 5.2.24 there are some improvements to the mount procedure that eosxd uses when it finds an eosxd is already running. That looks relevant, so I have some hope that you will get better behaviour with 5.2.24. Would it be possible for you to try running with that version?

**Comment by alexcern** [comment]
Thanks for the info @smithdh, the problem just happened again in webeos prod and this is indeed consistent with a problem in the mount procedure that eosxd uses when it finds an eosxd is already running.  

eosxd processes while problem was present:

```
root        9070  7.1  0.2 809952 72460 ?        S<sl Jul17 755:46 /usr/bin/eosxd /eos/user -o rw,fsname=user -oautofs
root        9071  0.0  0.0  75744 21600 ?        S    Jul17   0:00 /usr/bin/eosxd /eos/user -o rw,fsname=user -oautofs
root        9074  1.6  0.2 756192 68568 ?        S<sl Jul17 176:00 /usr/bin/eosxd /eos/project -o rw,fsname=project -oautofs
root        9077  0.0  0.0  75744 21600 ?        S    Jul17   0:00 /usr/bin/eosxd /eos/project -o rw,fsname=project -oautofs
root        9236  4.0  1.7 2430600 522980 ?      S<sl Jul17 433:08 /usr/bin/eosxd /eos/project-i02 -o rw,fsname=project-i02 -oautofs
root        9237  1.0  0.7 2776200 226600 ?      S<sl Jul17 106:02 /usr/bin/eosxd /eos/home-i00 -o rw,fsname=home-i00 -oautofs
root        9238  0.0  0.0  75744 21764 ?        S    Jul17   0:00 /usr/bin/eosxd /eos/project-i02 -o rw,fsname=project-i02 -oautofs
root        9239  0.0  0.0  75744 21700 ?        S    Jul17   0:00 /usr/bin/eosxd /eos/home-i00 -o rw,fsname=home-i00 -oautofs
root        9502  2.1  1.4 2462920 441412 ?      S<sl Jul17 227:49 /usr/bin/eosxd /eos/home-i02 -o rw,fsname=home-i02 -oautofs
root        9503  0.0  0.0  75744 21520 ?        S    Jul17   0:00 /usr/bin/eosxd /eos/home-i02 -o rw,fsname=home-i02 -oautofs
root       44750  0.4  0.4 1305728 132896 ?      S<sl Jul17  43:13 /usr/bin/eosxd /eos/web -o rw,fsname=web -oautofs
root       44751  0.0  0.0  75744 21716 ?        S    Jul17   0:00 /usr/bin/eosxd /eos/web -o rw,fsname=web -oautofs
root       44843  2.9  1.3 2206344 394588 ?      S<sl Jul17 307:29 /usr/bin/eosxd /eos/home-i01 -o rw,fsname=home-i01 -oautofs
root       44845  0.0  0.0  75744 21736 ?        S    Jul17   0:00 /usr/bin/eosxd /eos/home-i01 -o rw,fsname=home-i01 -oautofs
root     3366557  0.5  0.2 669020 69552 ?        S<sl 15:23   0:22 /usr/bin/eosxd /eos/project-i01 -o rw,fsname=project-i01 -oautofs
root     3366559  0.0  0.0  75744 21584 ?        S    15:23   0:00 /usr/bin/eosxd /eos/project-i01 -o rw,fsname=project-i01 -oautofs
root     3368259  0.2  0.2 655704 63132 ?        S<sl 15:23   0:09 /usr/bin/eosxd /eos/theory -o rw,fsname=theory -oautofs
root     3368261  0.0  0.0  75744 21464 ?        S    15:23   0:00 /usr/bin/eosxd /eos/theory -o rw,fsname=theory -oautofs
root     3723321  0.0  0.0   3968  2440 ?        S    16:33   0:00 /usr/bin/mount -t eosx -o fsname=project-i01 eosxd /eos/project-i01
root     3723322  0.0  0.0   4244  3288 ?        S    16:33   0:00 /bin/bash /sbin/mount.eosx eosxd /eos/project-i01 -o rw,fsname=project-i01
root     3723323  0.2  0.1  75744 39612 ?        S    16:33   0:00 /usr/bin/eosxd /eos/project-i01 -o rw,fsname=project-i01 -oautofs
root     3723327  0.0  0.0   3968  2396 ?        S    16:33   0:00 /usr/bin/mount -t eosx -o fsname=theory eosxd /eos/theory
root     3723328  0.0  0.0   4244  3296 ?        S    16:33   0:00 /bin/bash /sbin/mount.eosx eosxd /eos/theory -o rw,fsname=theory
root     3723329  0.6  0.1  75744 39468 ?        S    16:33   0:00 /usr/bin/eosxd /eos/theory -o rw,fsname=theory -oautofs
```

the ` /eos/theory` and `/eos/project-i01` have old processes started at 15:23 with no corresponding mount point left. We see new `eosxd` at 16:33 invoked to mount them again but somehow not succeeding. Only after killing the old processes did they manage to re-mount ` /eos/theory` and `/eos/project-i01`.

We can try to deploy 5.2.24 on webeos-prod since we don't know how to reproduce the issue in a dev environment.

5.2.22 is the latest release available in eosxd-csi at the moment (since https://github.com/cern-eos/eosxd-csi/pull/10) so we'd need a new release first in order to test it

**Comment by alexcern** [comment]
I took a look to https://gitlab.cern.ch/dss/eos/-/commit/9166033f2bea6f9328a21307f69ac665bc5d416c and I'm worried the situation will be worse with 5.2.24... Here's my understanding:

So we have a situation where automount calls the fusex mounter while an old eosxd process still holds `lockB` for that EOS instance

In 5.2.22 the mounter waits 5s for lockB then gives up and returns successfully without doing anything. This is exactly what we see in automounter logs.

In 5.2.24, IIUC we will take [this code path](https://gitlab.cern.ch/dss/eos/-/commit/9166033f2bea6f9328a21307f69ac665bc5d416c#d042467c7a38927e3680b670f0591d38f2787dba_212_135) and `stat` the path that automount invoked us to mount. But since the mount at this path is in progress, this `stat` call is going to block in automount until the ongoing mount operation completes, which now will never happen since the mount operation is blocked in the `stat` call.

What do you think @smithdh ?

**Comment by smithdh** [comment]
Hi @alexcern 
The stat at the mountpoint does't re-trigger autofs to mount: If you'd like to test this code path in a test environment you could try this:

In one shell change working directory into an autofs path, e.g. "cd /eos/instance/path". Then lazy unmount the mount point, e.g. "umount -l /eos/instance". The eosxd will keep running, since the shell's process is keeping the filesystem busy. (e.g. you can continue to work with the filesystem from that shell). But the filesystem is detached and doesn't appear to be mounted. If you again trigger the autofs (from the first or another) shell by accessing under /eos/instance it will trigger the code path you highlighted.

**Comment by alexcern** [comment]
Solved with #21
